les/lespay/server/balance.go

```diff
@@ -0,0 +1,555 @@
+	bt                               *BalanceTracker
+	lock                             sync.RWMutex
+	node                             *enode.Node
+	connAddress                      string
+	active                           bool
+	priority                         bool
+	capacity                         uint64
+	balance                          balance
+	posFactor, negFactor             PriceFactors
+	sumReqCost                       uint64
+	lastUpdate, nextUpdate, initTime mclock.AbsTime
+	updateEvent                      mclock.Timer
+	// since only a limited and fixed number of callbacks are needed, they are
+	// stored in a fixed size array ordered by priority threshold.
+	callbacks [balanceCallbackCount]balanceCallback
+	// callbackIndex maps balanceCallback constants to callbacks array indexes (-1 if not active)
+	callbackIndex [balanceCallbackCount]int
+	callbackCount int // number of active callbacks
+}
+
+// balance represents a pair of positive and negative balances
+type balance struct {
+	pos, neg utils.ExpiredValue
+}
+
+// balanceCallback represents a single callback that is activated when client priority
+// reaches the given threshold
+type balanceCallback struct {
+	id        int
+	threshold int64
+	callback  func()
+}
+
+// GetBalance returns the current positive and negative balance.
+func (n *NodeBalance) GetBalance() (uint64, uint64) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	now := n.bt.clock.Now()
+	n.updateBalance(now)
+	return n.balance.pos.Value(n.bt.posExp.LogOffset(now)), n.balance.neg.Value(n.bt.negExp.LogOffset(now))
+}
+
+// GetRawBalance returns the current positive and negative balance
+// but in the raw(expired value) format.
+func (n *NodeBalance) GetRawBalance() (utils.ExpiredValue, utils.ExpiredValue) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	now := n.bt.clock.Now()
+	n.updateBalance(now)
+	return n.balance.pos, n.balance.neg
+}
+
+// AddBalance adds the given amount to the positive balance and returns the balance
+// before and after the operation. Exceeding maxBalance results in an error (balance is
+// unchanged) while adding a negative amount higher than the current balance results in
+// zero balance.
+func (n *NodeBalance) AddBalance(amount int64) (uint64, uint64, error) {
+	var (
+		err         error
+		old, new    uint64
+		setPriority bool
+	)
+	n.bt.updateTotalBalance(n, func() bool {
+		now := n.bt.clock.Now()
+		n.updateBalance(now)
+
+		// Ensure the given amount is valid to apply.
+		offset := n.bt.posExp.LogOffset(now)
+		old = n.balance.pos.Value(offset)
+		if amount > 0 && (amount > maxBalance || old > maxBalance-uint64(amount)) {
+			err = errBalanceOverflow
+			return false
+		}
+
+		// Update the total positive balance counter.
+		n.balance.pos.Add(amount, offset)
+		setPriority = n.checkPriorityStatus()
+		n.checkCallbacks(now)
+		new = n.balance.pos.Value(offset)
+		n.storeBalance(true, false)
+		return true
+	})
+	if err != nil {
+		return old, old, err
+	}
+	if setPriority {
```

Comment:
Just realize the issue(maybe I'm wrong here). If we assign the `priority` flag and then release the lock, is it possible that the callback is invoked and then `priority` is removed but after that we execute `n.bt.ns.SetState(n.node, n.bt.PriorityFlag, nodestate.Flags{}, 0)`

Reply:
It is very unlikely to happen but sure, it was a bit racey. Now I changed the callback mechanism a bit in the last commit to get rid of the ugly goroutine call. This particular issue could be solved in an easier, more hacky way but I think it is nicer to make things deterministic and not use goroutines as an easy workaround.

Reply:
Yes, I understand the fix now.